Name: Darshil Dhandhukia

Roll No: C010

Subject: DENS

BTech EXTC

Aim: RSA Algorithm (Public Key Encryption)

In [ ]:
def gcd_numbers(p,q): #we are taking gcd of two numbers
  if q == 0: #to find remainder
    return p
  else:
    r = p % q
    return gcd_numbers(q,r)

In [ ]:
#this code is for multiplicative inverse.
def multi_inv(n,a): #n is for mod n value
  r1 = n #x column value initialization
  r2 = a #y column value initialization
  x1 = int(1)
  x2 = int(0)
  y1 = int(0)
  y2 = int(1)

  while r2>0: #till the time r2 is non-zero we need to keep dividing
    q = r1 // r2 #"//" is for floor division, "q" is for finding the quotient
    r = r1 - (q*r2) #for remainder.
    r1 = r2 #from here the next step starts. here we assign r2 value to r1 and the new remainder value to r2.
    r2 = r
    x = x1 - (q*x2)
    x1 = x2
    x2 = x
    y = y1 - (q*y2)
    y1 = y2
    y2 = y

  if y1<0: #this if loop is used when we get a negative number as the multiplicative inverse.
    y1 = y1 % n
  return r1, y1

In [ ]:
p = int(input("Enter 1st prime number: "))
q = int(input("Enter 2nd prime number: "))

plain_text = int(input("Enter plaintext: "))

#gcd_numbers(p,q)

Enter 1st prime number: 13
Enter 2nd prime number: 19
Enter plaintext: 10


In [ ]:
n = p*q #euler totient
phi_n = (p-1)*(q-1)

In [ ]:
phi_n

216

In [ ]:
# Generate encryption key in range 1<e<Pn
key= []
for i in range(2,phi_n):
  gcd_given = gcd_numbers(phi_n,i)
  if gcd_given == 1:
    key.append(i)

print("All possible values of encryption key are: ", key)

All possible values of encryption key are:  [5, 7, 11, 13, 17, 19, 23, 25, 29, 31, 35, 37, 41, 43, 47, 49, 53, 55, 59, 61, 65, 67, 71, 73, 77, 79, 83, 85, 89, 91, 95, 97, 101, 103, 107, 109, 113, 115, 119, 121, 125, 127, 131, 133, 137, 139, 143, 145, 149, 151, 155, 157, 161, 163, 167, 169, 173, 175, 179, 181, 185, 187, 191, 193, 197, 199, 203, 205, 209, 211, 215]


In [ ]:
# Select an encryption key from the above list
e= int(input("Choose a value for e from the above set: "))

Choose a value for e from the above set: 13


In [ ]:
#Obtain inverse of encryption key in Z_Pn
r, d = multi_inv(phi_n,e)
if r==1:
  d=int(d)
  print("decryption key is: ", d)
else:
  print("Multiplicative inverse for the given encryption key does not exist. Choose a different encrytion key ")

decryption key is:  133


In [ ]:
cipher_text = plain_text ** e
cipher_text = cipher_text % n
print("Cipher text is: ", cipher_text)

Cipher text is:  127


In [ ]:
#decrypting the ciphertext
decrypted_pt = cipher_text ** d
decrypted_pt = decrypted_pt % n
print("Decrypted plain text is: ", decrypted_pt)

Decrypted plain text is:  10


**2nd example**

Below I have encrypted the text "HI" by converting it to integer values, that is, 8 and 9 respectively. 

In [ ]:
p = int(input("Enter 1st prime number: "))
q = int(input("Enter 2nd prime number: "))

plain_text = int(input("Enter plaintext: "))

Enter 1st prime number: 53
Enter 2nd prime number: 59
Enter plaintext: 89


In [ ]:
n = p*q #euler totient
phi_n = (p-1)*(q-1)

phi_n

3016

In [ ]:
# Generate encryption key in range 1<e<Pn
key= []
for i in range(2,phi_n):
  gcd_given = gcd_numbers(phi_n,i)
  if gcd_given == 1:
    key.append(i)

print("All possible values of encryption key are: ", key)

All possible values of encryption key are:  [3, 5, 7, 9, 11, 15, 17, 19, 21, 23, 25, 27, 31, 33, 35, 37, 41, 43, 45, 47, 49, 51, 53, 55, 57, 59, 61, 63, 67, 69, 71, 73, 75, 77, 79, 81, 83, 85, 89, 93, 95, 97, 99, 101, 103, 105, 107, 109, 111, 113, 115, 119, 121, 123, 125, 127, 129, 131, 133, 135, 137, 139, 141, 147, 149, 151, 153, 155, 157, 159, 161, 163, 165, 167, 171, 173, 175, 177, 179, 181, 183, 185, 187, 189, 191, 193, 197, 199, 201, 205, 207, 209, 211, 213, 215, 217, 219, 223, 225, 227, 229, 231, 233, 235, 237, 239, 241, 243, 245, 249, 251, 253, 255, 257, 259, 263, 265, 267, 269, 271, 275, 277, 279, 281, 283, 285, 287, 289, 291, 293, 295, 297, 301, 303, 305, 307, 309, 311, 313, 315, 317, 321, 323, 327, 329, 331, 333, 335, 337, 339, 341, 343, 345, 347, 349, 353, 355, 357, 359, 361, 363, 365, 367, 369, 371, 373, 375, 379, 381, 383, 385, 387, 389, 391, 393, 395, 397, 399, 401, 405, 407, 409, 411, 413, 415, 417, 419, 421, 423, 425, 427, 431, 433, 437, 439, 441, 443, 445, 447, 449, 45

In [ ]:
# Select an encryption key from the above list
e= int(input("Choose a value for e from the above set: "))

Choose a value for e from the above set: 3


In [ ]:
#Obtain inverse of encryption key in Z_Pn
r, d = multi_inv(phi_n,e)
if r==1:
  d=int(d)
  print("decryption key is: ", d)
else:
  print("Multiplicative inverse for the given encryption key does not exist. Choose a different encrytion key ")

decryption key is:  2011


In [ ]:
cipher_text = plain_text ** e
cipher_text = cipher_text % n
print("Cipher text is: ", cipher_text)

Cipher text is:  1394


In [ ]:
#decrypting the ciphertext
decrypted_pt = cipher_text ** d
decrypted_pt = decrypted_pt % n
print("Decrypted plain text is: ", decrypted_pt)

Decrypted plain text is:  89


##Conclusion:
For the given plaintext (10) and public key(13, 216), the ciphertext using RSA algorithm is 127.

Ciphertext was decrypted using private key(13, 216), such that the decrypted message is the same as original message.

If chosen public key, which should be relatively prime, is not relatively prime to phi(n), then RSA algorithm will not decrypt the ciphertext to the actual message